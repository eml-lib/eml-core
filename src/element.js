import flatten from './helpers/array-flatten';

const reservedProps = {
	context: true
};

export const createElement = (type, props, ...children) => {
	const filteredProps = {};

	if (props) {
		Object.entries(props).forEach(([name, value]) => {
			if (!reservedProps.hasOwnProperty(name)) {
				filteredProps[name] = value;
			}
		});
	}

	if (type && type.defaultProps) {
		Object.entries(type.defaultProps).forEach(([name, value]) => {
			if (filteredProps[name] === undefined) {
				filteredProps[name] = value;
			}
		});
	}

	return {
		_isElement: true,
		type,
		props: {
			...filteredProps,
			children: children ? flatten(children) : []
		}
	};
};

export const cloneElement = (element, props, ...children) => {
	if (element === null || element === undefined) {
		throw new Error(`React.cloneElement(...): The argument must be a React element, but you passed ${element}`);
	}

	return createElement(element.type, );
};

export const isElement = object => (
	typeof object === 'object'
	&& object !== null
	&& object._isElement === true
);