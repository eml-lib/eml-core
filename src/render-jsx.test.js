import renderHtml from './render-html';
import renderJsx from './render-jsx';
import { createElement } from './element';
import Component from './component';
import Fragment from './fragment';

// import { inspect } from 'util';
// function log(object) {
//     return console.log(inspect(object, {
//         colors: true,
//         depth: Infinity
//     }));
// }

const B = props => {
	return (
		<div className="b">
			1
		</div>
	);
};
B.css = {
	'.b': { border: '1px solid green' }
};
B.styles = () => ({
	b: 2
});


class A extends Component {
	render() {
		const { size } = this.props;

		return (
			<div className="block">
				{ [1, 2, 3].map(i => (
					<div className="block__body">
						<B />
					</div>
				)) }
			</div>
		);
	}

	static css = {
		'.block': {
			border: '1px solid red'
		}
	};

	static styles = ({ size }) => ({
		'@media (max-width: 600px)': {
			'.block_body': {
				width: 100 / size + 'px'
			}
		}
	});

	// static css = {
	// 	a: { border: '1px solid red' },
	// 	a_css: { left: 0, top: 0 }
	// };
    //
	// static styles = () => ({
	// 	a: 1
	// });
}

// const { css, styles, html } = ;
//
// log({
// 	css,
// 	styles,
// 	html
// });

it('Self-closing elements', () => {
	expect(
		renderJsx(
			<Fragment>1</Fragment>
		)
	).toEqual({});
});