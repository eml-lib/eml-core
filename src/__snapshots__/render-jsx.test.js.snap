// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrays array in arrays 1`] = `
Array [
  "1",
]
`;

exports[`Arrays arrays in arrays 1`] = `
Array [
  "1",
  "2",
  "3",
]
`;

exports[`Element attributes 1`] = `
Object {
  "attrs": Object {
    "cellpadding": 0,
    "cellspacing": 0,
  },
  "children": Array [
    Object {
      "attrs": Object {},
      "children": Array [
        Object {
          "attrs": Object {
            "colspan": 2,
            "rowspan": 3,
          },
          "children": Array [],
          "tagName": "td",
        },
      ],
      "tagName": "tr",
    },
  ],
  "tagName": "table",
}
`;

exports[`Element one level 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    "1",
  ],
  "tagName": "div",
}
`;

exports[`Element two levels 1`] = `
Object {
  "attrs": Object {},
  "children": Array [
    Object {
      "attrs": Object {},
      "children": Array [
        "1",
      ],
      "tagName": "div",
    },
    Object {
      "attrs": Object {},
      "children": Array [
        "2",
      ],
      "tagName": "div",
    },
  ],
  "tagName": "div",
}
`;

exports[`Fragment 2 levels 1`] = `
Array [
  "1",
  "2",
]
`;

exports[`Fragment deep levels 1`] = `
Array [
  "1",
  "2",
  "3",
  "4",
]
`;

exports[`Fragment flat 1`] = `
Array [
  "1",
]
`;

exports[`Mixed fragments with arrays 1`] = `
Array [
  Object {
    "attrs": Object {},
    "children": Array [
      "1",
      "2",
    ],
    "tagName": "div",
  },
  Object {
    "attrs": Object {},
    "children": Array [
      "3",
      "4",
    ],
    "tagName": "div",
  },
  "[[5]]",
]
`;
